File: src/main/java/net/guizhanss/guizhancraft/GuizhanCraft.java
package net.guizhanss.guizhancraft;

import java.io.File;
import java.lang.reflect.Method;
import java.util.logging.Level;

import javax.annotation.Nonnull;
import javax.annotation.ParametersAreNonnullByDefault;

import org.bukkit.configuration.Configuration;
import org.bukkit.plugin.Plugin;

import io.github.thebusybiscuit.slimefun4.libraries.dough.updater.BlobBuildUpdater;

import net.guizhanss.guizhancraft.core.services.LocalizationService;
import net.guizhanss.guizhancraft.implementation.setup.CommandSetup;
import net.guizhanss.guizhancraft.implementation.setup.ItemSetup;
import net.guizhanss.guizhancraft.implementation.setup.ResearchSetup;
import net.guizhanss.guizhancraft.integrations.IntegrationsManager;
import net.guizhanss.guizhanlib.slimefun.addon.AbstractAddon;
import net.guizhanss.guizhanlib.slimefun.addon.AddonConfig;
import net.guizhanss.guizhanlib.updater.GuizhanBuildsUpdater;

public final class GuizhanCraft extends AbstractAddon {

    private static final String DEFAULT_LANG = "en-US";
    private static final int CONFIG_VERSION = 2;

    private boolean isDebugEnabled = false;

    // Services
    private LocalizationService localization;

    // Managers
    private IntegrationsManager integrations;

    public GuizhanCraft() {
        super("ybw0014", "GuizhanCraft", "master", "auto-update");
    }

    @Nonnull
    private static GuizhanCraft inst() {
        return getInstance();
    }

    @Nonnull
    public static LocalizationService getLocalization() {
        return inst().localization;
    }

    @Nonnull
    public static IntegrationsManager getIntegrationManager() {
        return inst().integrations;
    }

    @ParametersAreNonnullByDefault
    public static void debug(String message, Object... args) {
        if (inst().isDebugEnabled) {
            GuizhanCraft.log(Level.WARNING, "[DEBUG] " + message, args);
        }
    }

    @Override
    protected void enable() {
        sendConsole("&a==================");
        sendConsole("&a   GuizhanCraft   ");
        sendConsole("&a    by ybw0014    ");
        sendConsole("&a==================");

        // Config
        AddonConfig config = getAddonConfig();
        if (config.getInt("version", 1) < CONFIG_VERSION) {
            Configuration defaultConfig = config.getDefaults();
            for (String key : defaultConfig.getKeys(true)) {
                if (!config.contains(key)) {
                    config.set(key, defaultConfig.get(key));
                }
            }
            config.set("version", CONFIG_VERSION);
            config.save();
        }

        // Debug
        if (config.getBoolean("debug")) {
            isDebugEnabled = true;
        }

        // Localization
        String lang = config.getString("lang", DEFAULT_LANG);
        localization = new LocalizationService(this);
        localization.addLanguage(lang);
        if (!lang.equals(DEFAULT_LANG)) {
            localization.addLanguage(DEFAULT_LANG);
        }
        localization.sendConsole("loaded_language", lang);

        // Items
        localization.sendConsole("loading_items");
        ItemSetup.setup();

        // Commands
        localization.sendConsole("loading_commands");
        CommandSetup.setup();

        // Researches
        if (getConfig().getBoolean("enable-researches", true)) {
            localization.sendConsole("loading_researches");
            ResearchSetup.setup();
        }

        // Integrations
        localization.sendConsole("loading_integrations");
        integrations = new IntegrationsManager(this);
        integrations.start();
    }

    @Override
    protected void disable() {
        // Nothing to do yet
    }

    @Override
    protected void autoUpdate() {
        if (getPluginVersion().startsWith("Dev")) {
            new BlobBuildUpdater(this, getFile(), getGithubRepo()).start();
        } else if (getPluginVersion().startsWith("Build")) {
            try {
                // use updater in lib plugin

                // this little trick because maven shade plugin will change strings
                char[] pluginPackage = {
                    'n', 'e', 't', '.', 'g', 'u', 'i', 'z', 'h', 'a', 'n', 's', 's', '.',
                    'g', 'u', 'i', 'z', 'h', 'a', 'n', 'l', 'i', 'b', 'p', 'l', 'u', 'g', 'i', 'n'
                };
                Class<?> clazz = Class.forName(new String(pluginPackage) + ".updater.GuizhanUpdater");
                Method updaterStart = clazz.getDeclaredMethod("start", Plugin.class, File.class, String.class, String.class, String.class);
                updaterStart.invoke(null, this, getFile(), getGithubUser(), getGithubRepo(), getGithubBranch());
            } catch (Exception ignored) {
                // use updater in lib
                GuizhanBuildsUpdater.start(this, getFile(), getGithubUser(), getGithubRepo(), getGithubBranch());
            }
        }
    }
}

File: src/main/kotlin/net/guizhanss/guizhancraft/GuizhanCraft.kt
package net.guizhanss.guizhancraft

import io.github.thebusybiscuit.slimefun4.libraries.dough.updater.BlobBuildUpdater
import net.byteflux.libby.Library
import net.guizhanss.guizhancraft.core.services.ConfigService
import net.guizhanss.guizhancraft.core.services.IntegrationService
import net.guizhanss.guizhancraft.core.services.LocalizationService
import net.guizhanss.guizhancraft.implementation.commands.SfSpawnerCommand
import net.guizhanss.guizhancraft.implementation.commands.UnloadChunkCommand
import net.guizhanss.guizhancraft.implementation.groups.GCItemGroups
import net.guizhanss.guizhancraft.implementation.items.GCItems
import net.guizhanss.guizhancraft.implementation.setup.ResearchSetup
import net.guizhanss.guizhanlib.libraries.BukkitLibraryManager
import net.guizhanss.guizhanlib.slimefun.addon.AbstractAddon
import net.guizhanss.guizhanlib.updater.GuizhanBuildsUpdater
import org.bstats.bukkit.Metrics
import org.bukkit.Bukkit
import org.bukkit.plugin.Plugin
import java.io.File
import java.util.logging.Level

class GuizhanCraft : AbstractAddon(
    GITHUB_USER, GITHUB_REPO, GITHUB_BRANCH, AUTO_UPDATE_KEY
) {

    override fun load() {
        // check if there is central repo prop defined
        val centralRepo = System.getProperty("centralRepository") ?: "https://repo1.maven.org/maven2/"

        logger.info("Loading libraries, please wait...")
        logger.info("If you stuck here for a long time, try to specify a mirror repository:")
        logger.info("Add -DcentralRepository=<url> to the JVM arguments.")

        // download libs
        val manager = BukkitLibraryManager(this)
        manager.addRepository(centralRepo)
        manager.loadLibrary(
            Library.builder().groupId("org.jetbrains.kotlin").artifactId("kotlin-stdlib").version("2.1.20").build()
        )
        manager.loadLibrary(
            Library.builder().groupId("org.jetbrains.kotlin").artifactId("kotlin-reflect").version("2.1.20").build()
        )

        logger.info("Loaded all required libraries.")
    }

    override fun enable() {
        instance = this

        // config
        configService = ConfigService(this)
        debug("Debug mode is enabled.")

        // localization
        log(Level.INFO, "Loading language...")
        val lang = configService.lang
        localization = LocalizationService(this, file)
        localization.idPrefix = "GZ_"
        localization.addLanguage(lang)
        if (lang != DEFAULT_LANG) {
            localization.addLanguage(DEFAULT_LANG)
        }
        log(Level.INFO, "Loaded language {0}.", lang)

        // integrations
        integrationService = IntegrationService(this)

        // item groups setup
        GCItemGroups

        // item setup
        GCItems

        // researches setup
        if (configService.enableResearches) {
            ResearchSetup
        }

        // commands setup
        setupCommands()

        // Metrics setup
        setupMetrics()
    }

    override fun disable() {
        Bukkit.getScheduler().cancelTasks(this)
    }

    override fun autoUpdate() {
        if (pluginVersion.startsWith("Dev")) {
            BlobBuildUpdater(this, file, githubRepo).start()
        } else if (pluginVersion.startsWith("Build")) {
            try {
                // use updater in lib plugin
                val clazz = Class.forName("net.guizhanss.minecraft.guizhanlib.updater.GuizhanUpdater")
                val updaterStart = clazz.getDeclaredMethod(
                    "start",
                    Plugin::class.java,
                    File::class.java,
                    String::class.java,
                    String::class.java,
                    String::class.java
                )
                updaterStart.invoke(null, this, file, githubUser, githubRepo, githubBranch)
            } catch (_: Exception) {
                // use updater in lib
                GuizhanBuildsUpdater.start(this, file, githubUser, githubRepo, githubBranch)
            }
        }
    }

    private fun setupCommands() {
        SfSpawnerCommand(getPluginCommand("sfspawner")).register()
        UnloadChunkCommand(getPluginCommand("unloadchunk")).register()
    }

    private fun setupMetrics() {
        Metrics(this, 14629)
    }

    companion object {

        private const val GITHUB_USER = "ybw0014"
        private const val GITHUB_REPO = "GuizhanCraft"
        private const val GITHUB_BRANCH = "master"
        private const val AUTO_UPDATE_KEY = "auto-update"
        const val DEFAULT_LANG = "en-US"

        lateinit var instance: GuizhanCraft
            private set
        lateinit var configService: ConfigService
            private set
        lateinit var localization: LocalizationService
            private set
        lateinit var integrationService: IntegrationService
            private set

        fun scheduler() = getScheduler()

        fun log(level: Level, message: String) {
            instance.logger.log(level, message)
        }

        fun log(level: Level, ex: Throwable, message: String) {
            instance.logger.log(level, ex) { message }
        }

        fun debug(message: String) {
            if (!Companion::configService.isInitialized || !configService.debug) return
            log(Level.INFO, "[DEBUG] $message")
        }
    }
}
